add keepalived vrrp track_process support w/tests

jontow opened the pull request:
<!--
Thank you for contributing to this project!

- This project has a Contributor Code of Conduct: https://voxpupuli.org/coc/
- Please check that here is no existing issue or PR that addresses your problem.
- Our vulnerabilities reporting process is at https://voxpupuli.org/security/

-->
#### Pull Request (PR) description
keepalived 2.0.11 adds support for "track_process", which uses a process event monitoring system to avoid the overhead of pgrep/pkill when using track_script which has to crawl /proc for info.  This PR adds support for configuring the new track_process feature, as well as adding some basic unit tests for it.  Closely modeled on existing track_script support.

#### This Pull Request (PR) fixes the following issues
n/a

Dan33l commented on manifests/init.pp:
```diff
@@ -37,6 +38,7 @@
   validate_string($service_name)
   validate_hash($vrrp_instance)
   validate_hash($vrrp_script)
+  validate_hash($vrrp_track_process)
```
The next release will use Puppet 4 and so old `validate` functions are deprecated.

A datatype need to be used instead.
Details about this work in #163 .

Dan33l commented on spec/defines/keepalived_vrrp_track_process_spec.rb:
```diff
@@ -0,0 +1,104 @@
+require 'spec_helper'
+
+describe 'keepalived::vrrp::track_process', type: :define do
+  let(:facts) { debian_facts }
```
Please use https://github.com/mcanevet/rspec-puppet-facts.

alexjfisher replied:
This would be preferable.  Converting the rest of the module is on my todo list.

Dan33l commented on templates/track_process.erb:
```diff
@@ -0,0 +1,15 @@
+vrrp_track_process <%= @name %> {
```
epp template is welcome.

alexjfisher replied:
@Dan33l is right that this is often preferred but I'm also fine with `erb` (especially given all the existing templates are erb).

Dan33l commented on templates/vrrp_instance.erb:
```diff
@@ -86,6 +86,14 @@ vrrp_instance <%= @_name %> {
   }
```
epp template is welcome

Dan33l commented on manifests/config.pp:
```diff
@@ -49,6 +49,7 @@
 
   create_resources(keepalived::vrrp::instance, $::keepalived::vrrp_instance)
   create_resources(keepalived::vrrp::script, $::keepalived::vrrp_script)
+  create_resources(keepalived::vrrp::vrrp_track_process, $::keepalived::vrrp_track_process)
```
old fashioned method :-D

Are you able to use `.each` instead ?

alexjfisher replied:
@Dan33l I'll open a separate PR removing all the other `create_resources`.  This should clarify the preferred style.

alexjfisher replied:
This was merged in https://github.com/voxpupuli/puppet-keepalived/pull/165

Dan33l commented on manifests/vrrp/track_process.pp:
```diff
@@ -0,0 +1,39 @@
+#             Default: undef
+#
+define keepalived::vrrp::track_process (
+  $proc_name   = undef,
```
Data types are now present, please use them.

alexjfisher commented on manifests/vrrp/track_process.pp:
```diff
@@ -0,0 +1,39 @@
+  $delay       = undef,
+  $fullcommand = undef
+) {
+  if ! $proc_name {
```
Could you make the parameter non optional instead (remove its default)?  I'm fixing this for script.pp in https://github.com/voxpupuli/puppet-keepalived/pull/163

The odd construct was left over from https://github.com/voxpupuli/puppet-keepalived/commit/3a529bc7310782f55b03774d1c4e3a044d86549a#diff-5b0dbec169818e08fe4324e5386753c6L10

alexjfisher commented on manifests/vrrp/track_process.pp:
```diff
@@ -0,0 +1,39 @@
+define keepalived::vrrp::track_process (
+  $proc_name   = undef,
+  $weight      = undef,
+  $quorum      = '1',
```
Please make this an Integer instead of a String.
Probably `Integer[1] $quorum = 1,`

jontow commented:
Unfortunately, collaborative git is not my strongest game and it appears I completely blew it on the rebase w/merge conflicts.  I'm going to have to do the annoying thing and close this PR and open a new one addressing the feedback from above.

alexjfisher commented:
@jontow No worries. We could have probably helped you out.  If you do need any assistance, please drop by on Freenode IRC `#voxpupuli` channel or the Puppet Community slack channel of the same name.
